tests/checkers/unittest_unicode/bad_chars.py

```diff
@@ -0,0 +1,255 @@
+            )
+            # Only utf8 can drop invalid lines
+            if codec_and_msg[0].startswith("utf") or not suffix[0]
+        ],
+    )
+    def test_find_bad_chars_without_nul(
+        self,
+        tmp_path: Path,
+        codec_and_msg: Tuple[str, Tuple[pylint.testutils.MessageTest]],
+        line_ending: str,
+        add_invalid_bytes: bytes,
+    ):
+        """All combinations of bad characters that are accepted by Python at the moment
+        are tested in all possible combinations of
+          - line ending
+          - encoding
+          - including not encode-able byte (or not)
+        """
+        codec, start_msg = codec_and_msg
+
+        def encode_without_bom(string, encoding):
+            return pylint.checkers.unicode._encode_without_bom(string, encoding)
+
+        # All lines contain a not extra checked invalid character
+        lines = (
+            "# Example File containing bad ASCII",
+            "# invalid char backspace: \b",
+            "# Bad carriage-return \r # not at the end",
+            "# Invalid char sub: \x1A",
+            "# Invalid char esc: \x1B",
+        )
+        byte_suffix = b""
+        if add_invalid_bytes:
+            if codec == "utf-8":
+                byte_suffix = b"\x80abc"
+            elif codec == "utf-16":
+                byte_suffix = b"\n"  # Generates Truncated Data
+            else:
+                byte_suffix = b"\xc3\x28 "
+            byte_suffix = encode_without_bom(" foobar ", codec) + byte_suffix
+
+        line_ending_encoded = encode_without_bom(line_ending, codec)
+
+        # Start content with BOM / codec definition and two empty lines
+        content = f"# coding: {codec} \n # \n ".encode(codec)
+        start_lines = 2
+
+        # Generate context with the given codec and line ending
+        for lineno, line in enumerate(lines):
+            byte_line = encode_without_bom(line, codec)
+            if byte_suffix:
+                byte_line += byte_suffix
+            byte_line += line_ending_encoded
+            content += byte_line
+
+            # Directly test the generated content
+            if not add_invalid_bytes:
+                # Test that the content is correct and gives no errors
+                try:
+                    byte_line.decode(codec, "strict")
+                except UnicodeDecodeError as e:
+                    raise ValueError(
+                        f"Line {lineno} did raise unexpected error: {byte_line}\n{e}"
+                    ) from e
+            else:
+                try:
+                    # But if there was a byte_suffix we expect an error
+                    # because that is what we want to test for
+                    byte_line.decode(codec, "strict")
+                except UnicodeDecodeError:
+                    ...
+                else:
+                    raise ValueError(
+                        f"Line {lineno} did not raise decode error: {byte_line}"
+                    )
+
```

Comment:
```suggestion
  def test_(...):
```
This test function is big and seems to test something else entirely after this point.